core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -253,4 +396,110 @@ public Registration registerDispatchInterceptor(MessageDispatchInterceptor<Query
                             .map(queryHandler -> (MessageHandler<? super QueryMessage<?, ?>>) queryHandler)
                             .collect(Collectors.toList());
     }
+
+    private class SimpleQueryUpdateEmitter<U> implements QueryUpdateEmitter<U> {
+
+        private UpdateHandler<?, U> updateHandler;
+        private final AtomicReference<Runnable> registrationCanceledHandlerRef = new AtomicReference<>();
+        private volatile boolean active;
+        private final ReentrantReadWriteLock producerConsumerLock = new ReentrantReadWriteLock(true);
+        private final ReentrantLock initialLock;
+        private final Condition initialCondition;
+        private volatile boolean initial;
+
+        SimpleQueryUpdateEmitter(UpdateHandler<?, U> updateHandler, ReentrantLock initialLock,
+                                 Condition initialCondition) {
+            this.updateHandler = updateHandler;
+            this.initialLock = initialLock;
+            this.initialCondition = initialCondition;
+            this.active = true;
+            this.initial = true;
+        }
+
+        @Override
+        public boolean emit(U update) {
+            if (!waitForInitial()) {
+                return false;
+            }
+            ensureActive();
+            return readLockSafe(() -> updateHandler.onUpdate(update));
+        }
+
+        @Override
+        public boolean complete() {
+            if (!waitForInitial()) {
+                return false;
+            }
+            active = false;
+            return readLockSafe(() -> updateHandler.onCompleted());
+        }
+
+        @Override
+        public boolean error(Throwable error) {
+            if (!waitForInitial()) {
+                return false;
+            }
+            ensureActive();
+            return readLockSafe(() -> updateHandler.onError(error));
+        }
+
+        @Override
+        public void onRegistrationCanceled(Runnable r) {
+            registrationCanceledHandlerRef.set(r);
+        }
+
+        private void cancelRegistration() {
+            producerConsumerLock.writeLock().lock();
+            updateHandler = null;
+            invokeRegistrationCanceledHandler();
+            producerConsumerLock.writeLock().unlock();
+        }
+
+        private boolean waitForInitial() {
+            if (initial) {
+                initialLock.lock();
+                try {
+                    initialCondition.await();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    return false;
+                } finally {
+                    initialInvoked();
+                    initialLock.unlock();
+                }
+            }
+            return true;
+        }
+
+        private void initialInvoked() {
+            this.initial = false;
+        }
+
+        private boolean readLockSafe(Runnable r) {
+            producerConsumerLock.readLock().lock();
```

Comment:
Hmm.... would be nice to be able to do this without locks, but I can't find a way to guarantee no more invocations happen after a `Registration.cancel()` without using them.